jerry-core/ecma/builtin-objects/ecma-builtin-date.cpp

```diff
@@ -38,19 +41,258 @@
  */
 
 /**
+ * Helper function to try to parse a part of a date string
+ *
+ * @return NaN if cannot read from string, ToNumber() otherwise
+ */
+static ecma_number_t
+ecma_date_parse_date_chars (lit_utf8_iterator_t *iter, /**< iterator of the utf8 string */
+                            uint32_t num_of_chars) /**< number of characters to read and convert */
+{
+  ecma_number_t ret_value = ecma_number_make_nan ();
+  lit_utf8_size_t copy_size = 0;
+  const lit_utf8_byte_t *str_start_p = iter->buf_p + iter->buf_pos.offset;
+
+  while (num_of_chars--)
+  {
+    if (lit_utf8_iterator_is_eos (iter))
+    {
+      copy_size = 0;
+      break;
+    }
+
+    copy_size += lit_get_unicode_char_size_by_utf8_first_byte (*iter->buf_p);
+    lit_utf8_iterator_incr (iter);
+  }
+
+  if (copy_size > 0)
+  {
+    ecma_string_t *str_p = ecma_new_ecma_string_from_utf8 (str_start_p, copy_size);
+    ret_value = ecma_string_to_number (str_p);
+    ecma_deref_ecma_string (str_p);
+  }
+
+  return ret_value;
+} /* ecma_date_parse_date_chars */
+
+/**
  * The Date object's 'parse' routine
  *
  * See also:
  *          ECMA-262 v5, 15.9.4.2
+ *          ECMA-262 v5, 15.9.1.15
  *
  * @return completion value
  *         Returned value must be freed with ecma_free_completion_value.
  */
 static ecma_completion_value_t
-ecma_builtin_date_parse (ecma_value_t this_arg, /**< this argument */
+ecma_builtin_date_parse (ecma_value_t this_arg __attr_unused___, /**< this argument */
                          ecma_value_t arg) /**< string */
 {
-  ECMA_BUILTIN_CP_UNIMPLEMENTED (this_arg, arg);
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+  ecma_number_t *date_num_p = ecma_alloc_number ();
+  *date_num_p = ecma_number_make_nan ();
+
+  /* Date Time String fromat (ECMA-262 v5, 15.9.1.15) */
+  ECMA_TRY_CATCH (date_str_value,
+                  ecma_op_to_string (arg),
+                  ret_value);
+
+  ecma_string_t *date_str_p = ecma_get_string_from_value (date_str_value);
+
+  lit_utf8_size_t date_str_size = ecma_string_get_size (date_str_p);
+  MEM_DEFINE_LOCAL_ARRAY (date_start_p, date_str_size, lit_utf8_byte_t);
+
+  ecma_string_to_utf8_string (date_str_p, date_start_p, (ssize_t) date_str_size);
+  lit_utf8_iterator_t iter = lit_utf8_iterator_create (date_start_p, date_str_size);
+
+  /* 1. read year */
+  ecma_number_t year = ecma_date_parse_date_chars (&iter, 4);
+
+  if (!ecma_number_is_nan (year)
+      && year >= 0)
+  {
+    ecma_number_t month = ECMA_NUMBER_ONE;
+    ecma_number_t day = ECMA_NUMBER_ONE;
+    ecma_number_t time = ECMA_NUMBER_ZERO;
+
+    /* 2. read month if any */
+    if (!lit_utf8_iterator_is_eos (&iter)
```

Comment:
Do we really need to check the `eos` in the `if` cases? Isn't the `peek_next` enough?